File: vaaman_gst/vaaman_gst/match_status.py
import frappe


@frappe.whitelist()
def set_match_status(purchase_invoice):
    if not purchase_invoice:
        return ""

    match_status = frappe.db.get_value(
        "GST Inward Supply",
        {"link_name": purchase_invoice},
        "match_status",
    ) or ""

    current = frappe.db.get_value(
        "Purchase Invoice", purchase_invoice, "custom_match_status"
    )

    if current != match_status:
        frappe.db.set_value(
            "Purchase Invoice",
            purchase_invoice,
            "custom_match_status",
            match_status,
            update_modified=False,
        )

    return match_status

File: vaaman_gst/public/js/match_status.js
frappe.ui.form.on('Purchase Invoice', {
    refresh: function (frm) {
        if (!frm.doc.name || frm.is_new()) return;

        frappe.call({
            method: 'vaaman_gst.vaaman_gst.match_status.set_match_status',
            args: {
                purchase_invoice: frm.doc.name,
            },
            callback: function (r) {
                if (r.message === undefined) return;
                if (frm.doc.custom_match_status === r.message) return;

                frm.doc.custom_match_status = r.message;
                frm.refresh_field('custom_match_status');
            },
        });
    },
});